include/swift/Demangling/TypeDecoder.h

```diff
@@ -330,6 +335,95 @@ getObjCClassOrProtocolName(NodePointer node) {
 }
 #endif
 
+template <typename BuiltType, typename BuiltRequirement,
+          typename BuiltLayoutConstraint, typename BuilderType>
+void decodeRequirement(
+    NodePointer node, std::vector<BuiltRequirement> &requirements,
+    BuilderType &Builder, std::function<BuiltType(NodePointer)> decodeTypeFn,
+    std::function<BuiltLayoutConstraint(LayoutConstraintKind)>
```

Comment:
Can decodeTypeFn, getLayoutConstraintFn and so on be methods on the BuilderType concept passed into the TypeDecoder template? Remember it's a usage of CRTP:

```
class Foo : TypeDecoder<Foo> {
  BuiltType decodeType(NodePointer) { ... }
  BuiltLayoutConstraint decodeLayoutConstraint(LayoutConstraintKind)...
  ...
}
```

Reply:
The Builder implementations don't inherit from TypeDecoder, but I still managed to make those functions part of the BuilderType concept.

Reply:
*And* it made me realize that I forgot to actually replace the original source of the decodeRequirement code with a call to decodeRequirement!